### PR TITLE
First try on box-ing

### PR DESCRIPTION
**Pull request by szepeviktor**
`box compile` will produce a phar :)

https://github.com/humbug/box

### CODE REVIEW

**Comment by szepeviktor** [comment]
:( Nos, it does not work, I don't know how to configure box.

**Comment by szepeviktor** [comment]
Fixed.

**Comment by pxgamer** [comment]
Have you thought about using compactors such as these?
- `KevinGH\Box\Compactor\Php`
- `KevinGH\Box\Compactor\PhpScoper`

Also, regarding the `files` array, I assume this was added to fix the errors that appear. 🤔 I feel like the Phar shouldn't include tests, so I wonder if there's a workaround for these errors. Unless I've missed something.

**Comment by szepeviktor** [comment]
Hello @pxgamer ! I am nooob in configuring box.

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.15/composer.json#L61-L76

**Comment by pxgamer** [comment]
Ah, thanks for letting me know about that. Odd that those are in the `tests` directory, but makes sense to me now. 👍

No problem about the config, compactors are just a suggestion. 🙂

**Comment by pxgamer** [comment]
@szepeviktor, apologies. I hadn't had a chance to test this myself. Looks like this project doesn't work well with the `PhpScoper` compactor. Might be best to exclude this to start.  :thinking:

For reference, this is the output below:
<details>
<summary>Error log</summary>

```
PHP Fatal error:  Interface '_HumbugBox5ad79dbb92d1\PHP_CodeSniffer\Sniffs\Sniff' not found in phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/vendor/object-calisthenics/phpcs-calisthenics-rules/src/ObjectCalisthenics/Sniffs/Classes/ForbiddenPublicPropertySniff.php on line 11
PHP Stack trace:
PHP   1. {main}() /home/pxgamer/projects/php/phpinsights/phpinsights.phar:0
PHP   2. require() /home/pxgamer/projects/php/phpinsights/phpinsights.phar:14
PHP   3. _HumbugBox5ad79dbb92d1\{closure:phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/phpinsights:9-29}() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/phpinsights:29
PHP   4. _HumbugBox5ad79dbb92d1\Symfony\Component\Console\Application->run() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/phpinsights:28
PHP   5. _HumbugBox5ad79dbb92d1\Symfony\Component\Console\Application->doRun() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/vendor/symfony/console/Application.php:101
PHP   6. _HumbugBox5ad79dbb92d1\Symfony\Component\Console\Application->doRunCommand() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/vendor/symfony/console/Application.php:190
PHP   7. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Application\Console\Commands\InvokableCommand->run() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/vendor/symfony/console/Application.php:586
PHP   8. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Application\Console\Commands\InvokableCommand->execute() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/vendor/symfony/console/Command/Command.php:127
PHP   9. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Application\Console\Commands\AnalyseCommand->__invoke() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Application/Console/Commands/InvokableCommand.php:21
PHP  10. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Application\Console\Analyser->analyse() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Application/Console/Commands/AnalyseCommand.php:33
PHP  11. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Domain\Insights\InsightCollectionFactory->get() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Application/Console/Analyser.php:18
PHP  12. array_map() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:45
PHP  13. _HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Domain\Insights\InsightCollectionFactory->_HumbugBox5ad79dbb92d1\NunoMaduro\PhpInsights\Domain\Insights\{closure:phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:40-45}() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:45
PHP  14. class_implements() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:41
PHP  15. spl_autoload_call() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:41
PHP  16. Composer\Autoload\ClassLoader->loadClass() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/src/Domain/Insights/InsightCollectionFactory.php:41
PHP  17. Composer\Autoload\includeFile() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/.box/vendor/composer/ClassLoader.php:322
PHP  18. include() phar:///home/pxgamer/projects/php/phpinsights/phpinsights.phar/.box/vendor/composer/ClassLoader.php:444
```

</details>

**Comment by nunomaduro** [comment]
Hey buddies, what is the current state of this pull request?

**Comment by szepeviktor** [comment]
Official donno.

**Comment by pxgamer** [comment]
@nunomaduro, I tested the compiled Phar on various repositories of my code and it seemed to be working fine. :+1:

However, when compiling with the latest code, I now get an issue which I'm not sure if it's related to Box or not.

**Comment by theofidry** [comment]
The config looks ok to me and I would recommend to compress the PHAR as well (`"compression": "GZ"`).

Other recommendations (which can be done in separate PRs):

- You can build & deploy the PHAR to attach it to the releases (you can check Box itself for the example)
- Adding a end-to-end test to check that the PHAR works as expected is a nice-to-have
- Adding compactors such as PhpScoper are a niceties but requires further tinkering and most definitely require end-to-end tests. The Json & Php compactors however should be relatively safe to use

**Comment by nunomaduro** [comment]
Just compiled with the current config and it's not working. Got an random error:
```
Cannot autowire service "Symplify\EasyCodingStandard\FixerRunner\Application\FixerFileProcessor": argument "$fixers" of method "__construct()" is type-hinted "array", you should configure its value explicitly.
```

It would be great have this tho.

**Comment by szepeviktor** [comment]
Try without the compactor!

**Comment by nunomaduro** [comment]
@theofidry Do you see any issue distributing `php-insights` just using the box compiled file? As I am doing here https://github.com/laravel-zero/installer/blob/stable/composer.json? Where basically people just get the binary in their vendor?

**Comment by nunomaduro** [comment]
PS: @szepeviktor Yes, it works. 👍

**Comment by theofidry** [comment]
>Where basically people just get the binary in their vendor?

Yeah that's fine, although I tend to recommend https://phar.io/ or a sort of installer script like indicated [there](https://github.com/box-project/box2#as-a-phar-recommended although there is yet to be a way to generate it from Box itself: https://github.com/humbug/box/issues/8

@szepeviktor it would be really nice to know how the compactor messed up there since it should only remove comments and unnecessary annotations

**Comment by szepeviktor** [comment]
It was just a tip. I am not into box.

File: box.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "files": [
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/AbstractFixerTestCase.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/AbstractIntegrationCaseFactory.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/AbstractIntegrationTestCase.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/Assert/AssertTokensTrait.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/IntegrationCase.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/IntegrationCaseFactory.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/IntegrationCaseFactoryInterface.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/Test/InternalIntegrationCaseFactory.php",
+        "vendor/friendsofphp/php-cs-fixer/tests/TestCase.php"
```
Comment (theofidry) [inline box.json]: can't it be simplified into:

```json
"directories: [
  "config",
  "vendor/friendsofphp/php-cs-fixer/tests/Test"
}
```

?

Comment (pxgamer) [reply box.json]: It looks like that would work 👍

Comment (szepeviktor) [reply box.json]: except  "vendor/friendsofphp/php-cs-fixer/tests/TestCase.php"

Comment (pxgamer) [reply box.json]: Ah. 🤔

Comment (theofidry) [reply box.json]: Does it matter though?

Comment (pxgamer) [reply box.json]: Probably as the testcases in `tests/Test` extend it.

**Comment by nunomaduro** [comment]
At the moment, this is not working. Let's wait until someone proposes a solutions for this.

**Pull request closed without merging**